File: baekjoon/gold/23288/23288.py
from collections import deque

def solution():
    answer = 0
    axis_arr = deque([6, 3, 1, 4])
    rest_arr = [5, 2]
    dir = 0
    N, M, K = map(int, input().split(" "))
    board = [list(map(int, input().split(" "))) for _ in range(N)]
    r, c = 0, 0
    dx, dy = [0, 1, 0, -1], [1, 0, -1, 0]
    score = [[-1 for _ in range(M)] for _ in range(N)]

    def dfs(r, c, t):
        ret = [(r, c)]
        score[r][c] = 1
        for i in range(4):
            rr, cc = r + dx[i], c + dy[i]
            if rr < 0 or N <= rr or cc < 0 or M <= cc or score[rr][cc] != -1 or board[rr][cc] != t:
                continue
            else:
                score[rr][cc] = 1
            ret.extend(dfs(rr, cc, t))
        return ret

    for i in range(N):
        for j in range(M):
            if score[i][j] == -1:
                same = dfs(i, j, board[i][j])
                multi = len(same)
                for x, y in same:
                    score[x][y] = multi * board[i][j]

    def convert_axis(axis_arr, rest_arr):
        axis_arr, rest_arr = deque([axis_arr[0], rest_arr[0], axis_arr[2], rest_arr[1]]), [axis_arr[1], axis_arr[3]]
        return [axis_arr, rest_arr]

    def rotate(dir, axis_arr, rest_arr):
        dir = (dir + 1) % 4
        return [dir] + convert_axis(axis_arr, rest_arr)

    def reverse_rotate(dir, axis_arr, rest_arr):
        dir = (dir + 3) % 4
        return [dir] + convert_axis(axis_arr, rest_arr)

    def reverse_dir(dir):
        dir = (dir + 2) % 4
        return dir

    def go(dir, axis_arr: deque, r, c):
        if dir < 2:
            axis_arr.append(axis_arr.popleft())
        else:
            axis_arr.appendleft(axis_arr.pop())
        r, c = r + dx[dir], c + dy[dir]
        return dir, axis_arr, r, c

    r, c, dir = 0, 0, 0

    for i in range(K):
        rr, cc = r + dx[dir], c + dy[dir]
        if rr < 0 or N <= rr or cc < 0 or M <= cc:
            dir = reverse_dir(dir)
        dir, axis_arr, r, c = go(dir, axis_arr, r, c)
        if axis_arr[0] < board[r][c]:
            dir, axis_arr, rest_arr = reverse_rotate(dir, axis_arr, rest_arr)
        elif axis_arr[0] > board[r][c]:
            dir, axis_arr, rest_arr = rotate(dir, axis_arr, rest_arr)
        answer += score[r][c]
    return answer
print(solution())

File: baekjoon/gold/23288/23288.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

var (
	reader  *bufio.Reader = bufio.NewReader(os.Stdin)
	writer  *bufio.Writer = bufio.NewWriter(os.Stdout)
	dx                    = []int{0, 1, 0, -1}
	dy                    = []int{1, 0, -1, 0}
	axisArr               = []int{6, 3, 1, 4}
	restArr               = []int{5, 2}
	score   [][]int
	board   [][]int
	N, M, K int
)

func solution() int {
	defer writer.Flush()
	answer := 0

	dir := 0

	fmt.Fscanf(reader, "%d %d %d\n", &N, &M, &K)

	board = make([][]int, N)
	score = make([][]int, N)
	for i := 0; i < N; i++ {
		board[i] = make([]int, M)
		score[i] = make([]int, M)
		for j := 0; j < M; j++ {
			fmt.Fscanf(reader, "%d ", &board[i][j])
			score[i][j] = -1
		}
	}

	for i := 0; i < N; i++ {
		for j := 0; j < M; j++ {
			if score[i][j] == -1 {
				same := dfs(i, j, board[i][j])
				multi := len(same)
				for _, v := range same {
					x, y := v[0], v[1]
					score[x][y] = multi * board[i][j]
				}
			}
		}
	}

	r, c, dir := 0, 0, 0

	for i := 0; i < K; i++ {
		rr, cc := r+dx[dir], c+dy[dir]
		if rr < 0 || rr >= N || cc < 0 || cc >= M {
			dir = reverseDir(dir)
		}
		dir, axisArr, r, c = goFunc(dir, axisArr, r, c)
		if axisArr[0] < board[r][c] {
			dir, axisArr, restArr = reverseRotate(dir, axisArr, restArr)
		} else if axisArr[0] > board[r][c] {
			dir, axisArr, restArr = rotate(dir, axisArr, restArr)
		}
		answer += score[r][c]
	}
	return answer
}

func main() {
	fmt.Println(solution())
}

func dfs(row, col, t int) [][2]int {
	ret := [][2]int{{row, col}}
	score[row][col] = 1
	for i := 0; i < 4; i++ {
		rr, cc := row+dx[i], col+dy[i]
		if rr < 0 || rr >= N || cc < 0 || cc >= M || score[rr][cc] != -1 || board[rr][cc] != t {
			continue
		}
		ret = append(ret, dfs(rr, cc, t)...)
	}
	return ret
}

func convertAxis(axisArr, restArr []int) {
	axisArr1, axisArr3 := axisArr[1], axisArr[3]
	restArr0, restArr1 := restArr[0], restArr[1]
	axisArr[1], axisArr[3] = restArr0, restArr1
	restArr[0], restArr[1] = axisArr1, axisArr3
}

func rotate(dir int, axisArr, restArr []int) (int, []int, []int) {
	dir = (dir + 1) % 4
	convertAxis(axisArr, restArr)
	return dir, axisArr, restArr
}

func reverseRotate(dir int, axisArr, restArr []int) (int, []int, []int) {
	dir = (dir + 3) % 4
	convertAxis(axisArr, restArr)
	return dir, axisArr, restArr
}

func reverseDir(dir int) int {
	return (dir + 2) % 4
}

func goFunc(dir int, axisArr []int, r, c int) (int, []int, int, int) {
	if dir < 2 {
		axisArr = append(axisArr[1:], axisArr[0])
	} else {
		axisArr = append([]int{axisArr[len(axisArr)-1]}, axisArr[:len(axisArr)-1]...)
	}
	r, c = r+dx[dir], c+dy[dir]
	return dir, axisArr, r, c
}
